Rewrite the assembly conflict resolution doc to describe current behavior

vitek-karas opened the pull request:
The doc was originally written as a proposal for a behavior change. So it's hard to read and determine the current behavior. Since the new behavior has been in place since 2.1, it's more valuable to have only the current version described.

vitek-karas committed:
Fix ordering of probe locations in assembly resolution doc

And refer to the detailed doc about this topic

noahfalk commented on docs/design/features/assembly-conflict-resolution.md:
```diff
@@ -17,19 +17,12 @@ Since each layer (and other probing locations) can contain the same assembly by
-*	`C:\\Users\\<user>\\.nuget\\packages`
-	`C:\\Program Files (x86)\\Microsoft SDKs\\NuGetPackagesFallback`
-*	`C:\\Program Files\\dotnet\\sdk\\NuGetFallbackFolder`
+1.  Framework directory(s) from higher to lower
```
The doc is no longer self-consistent unforetunately
In the section marked 'Algorithm' the text refers to app being after framework.
In the section marked 'Changes for 2.1+' it is refers back to this list with the expectation that framework is item 3 and app is item 4.

vitek-karas replied:
Good catch. I rewrote the doc almost fully to only describe the current behavior - it's not very useful to have a "proposal" doc around since some pieces of it were never implemented and it's hard to read to determine the current behavior.

vitek-karas committed:
Rewrite the document to describe only existing behavior

The doc was originally written as a proposal for a behavior change. So it's hard to read and determine the current behavior. Since the new behavior has been in place since 2.1, it's more valuable to have only the current version described.

noahfalk reviewed (approved):
Thanks for updating this!

elinor-fung commented on docs/design/features/assembly-conflict-resolution.md:
```diff
@@ -41,25 +32,18 @@ The order in which each layer's deps.json is processed is:
+1. ->If there's already a resolved entry with the same asset name, then check the assembly and file version of the new entry against the already resolved one. If the new entry is equal or lower, skip it. Otherwise remove the existing one and go to the "else" branch below
+1. ->Else (new asset, or replaced with higher version) probe for the actual asset file
+1. -->For each probing path except frameworks that are higher-level
+1. --->If the probing path is a framework, then check its `.deps.json` to see if it contains the exact package (by name and version). If it is, then use that location and end probing for this entry.
```
```suggestion
1. --->If the probing path is a framework, then check its `.deps.json` to see if it contains the exact package (by name and version). If it does, then use that location and end probing for this entry.
```

elinor-fung commented on docs/design/features/assembly-conflict-resolution.md:
```diff
@@ -41,25 +32,18 @@ The order in which each layer's deps.json is processed is:
-* For each probing path except frameworks that are higher-level
-
-The reason for this change is that a given framework should never have a dependency on a higher-level framework, and is expected to find deps assets in its layer or lower. This is also required so that the given framework can find its asset, and replace the higher-level asset (see next paragraph).
+Note that for an app both it's `.deps.json` as well as probing path comes before the framework's so mostly it will win. Also because the app will likely reference an OOB package that the framework doesn't (because a framework, at least Microsoft.NETCore.App, has its own metapackage and does not reference OOB packages), the framework probing path never matches up in step 4 for the app's `.deps.jso ...
```
```suggestion
Note that for an app both its `.deps.json` as well as probing path comes before the framework's `.deps.json`, so the app will usually win. Also because the app will likely reference an OOB package that the framework doesn't (because a framework, at least Microsoft.NETCore.App, has its own metapackage and does not reference OOB packages), the framework probing path never matches up in step 4 for the app's `.deps.json` package\assembly entry, so the "app wins".
```

elinor-fung replied:
And I think `step 4` -> `step 6`

elinor-fung commented on docs/design/features/assembly-conflict-resolution.md:
```diff
@@ -41,25 +32,18 @@ The order in which each layer's deps.json is processed is:
 
-Replace step 4 under **Algorithm** above with:
-* If the probing path is a framework, then check its deps.json to see if it contains the exact package (by name and version). If so check if a [minor] or [major] roll-forward occurred for this framework. If true (roll-forward), then check its deps.json to see if it contains a newer version of the assembly (by Assembly Version and then File Version if necessary) compared to a previously found deps ...
+The reason for only probing paths from equal or lower level framework in step 3 is that a given framework should never have a dependency on a higher-level framework, and is expected to find deps assets in its layer or lower. This is also required so that the given framework can find its asset, and replace the higher-level asset (see next paragraph).
```
Is `step 3` -> `step 5`?

elinor-fung commented on docs/design/features/assembly-conflict-resolution.md:
```diff
@@ -17,19 +15,12 @@ Since each layer (and other probing locations) can contain the same assembly by
 #### Probe Ordering
 The probing locations consists of:
 1.  Serving Location
```
```suggestion
1.  Servicing Location
```

Per #32272

vitek-karas committed:
Apply suggestions from code review

Co-Authored-By: Elinor Fung <47805090+elinor-fung@users.noreply.github.com>

vitek-karas committed:
Fix step numbers
